File: proj.android_studio/ee-x-app-lovin/src/main/java/com/ee/applovin/AppLovinInterstitialAdListener.kt
package com.ee.applovin

import com.applovin.sdk.AppLovinAd
import com.applovin.sdk.AppLovinAdClickListener
import com.applovin.sdk.AppLovinAdDisplayListener
import com.applovin.sdk.AppLovinAdLoadListener
import com.ee.core.IMessageBridge
import com.ee.core.Logger
import com.ee.core.internal.Thread
import java.util.concurrent.atomic.AtomicBoolean

internal class AppLovinInterstitialAdListener(
    private val _bridge: IMessageBridge,
    private val _isLoaded: AtomicBoolean)
    : AppLovinAdLoadListener
    , AppLovinAdDisplayListener
    , AppLovinAdClickListener {
    companion object {
        private val _logger = Logger(AppLovinInterstitialAdListener::class.java.name)

        private const val kPrefix = "AppLovinBridge"
        private const val kOnInterstitialAdLoaded = "${kPrefix}OnInterstitialAdLoaded"
        private const val kOnInterstitialAdFailedToLoad = "${kPrefix}OnInterstitialAdFailedToLoad"
        private const val kOnInterstitialAdClicked = "${kPrefix}OnInterstitialAdClicked"
        private const val kOnInterstitialAdClosed = "${kPrefix}OnInterstitialAdClosed"
    }

    override fun adReceived(ad: AppLovinAd) {
        _logger.info(this::adReceived.name)
        _isLoaded.set(true)
        _bridge.callCpp(kOnInterstitialAdLoaded)
    }

    override fun failedToReceiveAd(errorCode: Int) {
        _logger.info("${this::failedToReceiveAd.name}: code $errorCode")
        _bridge.callCpp(kOnInterstitialAdFailedToLoad, errorCode.toString())
    }

    override fun adDisplayed(ad: AppLovinAd) {
        _logger.info(this::adDisplayed.name)
        Thread.checkMainThread()
    }

    override fun adClicked(ad: AppLovinAd) {
        _logger.info(this::adClicked.name)
        Thread.checkMainThread()
        _bridge.callCpp(kOnInterstitialAdClicked)
    }

    override fun adHidden(ad: AppLovinAd) {
        _logger.info(this::adHidden.name)
        Thread.checkMainThread()
        _isLoaded.set(false)
        _bridge.callCpp(kOnInterstitialAdClosed)
    }
}

File: src/ee/app_lovin/private/AppLovinInterstitialAdListener.swift
//
//  AppLovinInterstitialAdListener.swift
//  Pods
//
//  Created by eps on 6/26/20.
//

import AppLovinSDK

private let kTag = "\(AppLovinInterstitialAdListener.self)"
private let kPrefix = "AppLovinBridge"
private let kOnInterstitialAdLoaded = "\(kPrefix)OnInterstitialAdLoaded"
private let kOnInterstitialAdFailedToLoad = "\(kPrefix)OnInterstitialAdFailedToLoad"
private let kOnInterstitialAdClicked = "\(kPrefix)OnInterstitialAdClicked"
private let kOnInterstitialAdClosed = "\(kPrefix)OnInterstitialAdClosed"

internal class AppLovinInterstitialAdListener:
    NSObject, ALAdLoadDelegate, ALAdDisplayDelegate {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private var _isLoaded = false

    init(_ bridge: IMessageBridge, _ logger: ILogger) {
        _bridge = bridge
        _logger = logger
        super.init()
    }

    var isLoaded: Bool {
        return _isLoaded
    }

    func adService(_ adService: ALAdService, didLoad ad: ALAd) {
        _logger.debug("\(kTag): \(#function)")
        _isLoaded = true
        _bridge.callCpp(kOnInterstitialAdLoaded)
    }

    func adService(_ adService: ALAdService, didFailToLoadAdWithError code: Int32) {
        _logger.debug("\(kTag): \(#function): code \(code)")
        _bridge.callCpp(kOnInterstitialAdFailedToLoad, "\(code)")
    }

    func ad(_ ad: ALAd, wasDisplayedIn view: UIView) {
        _logger.debug("\(kTag): \(#function)")
        _isLoaded = false
    }

    func ad(_ ad: ALAd, wasClickedIn view: UIView) {
        _logger.debug("\(kTag): \(#function)")
        _bridge.callCpp(kOnInterstitialAdClicked)
    }

    func ad(_ ad: ALAd, wasHiddenIn view: UIView) {
        _logger.debug("\(kTag): \(#function)")
        _bridge.callCpp(kOnInterstitialAdClosed)
    }
}
